### ISSUE TITLE
Modify function declarations to be more easily parsed by Visual Studio static analyzer

### ISSUE THREAD

**Issue by ianmackenzie** [ianmackenzie/opensolid#56]
The Visual Studio static analyzer fails to match class member function declarations to their definitions when the declaration involves a type such as `Matrix<double, iNumDimensions, iNumDimensions>`. It appears that the optional (omitted) `Matrix` template arguments are interfering with type-matching, as declaring a nested typedef such as `TransformationMatrixType` within the class and then using that in the function declaration/definition seems to fix the issue. Member functions should be updated to use this style where it does not hurt readability.

----------------------------------------
- Bitbucket: https://bitbucket.org/ianmackenzie/opensolid/issue/41
- Originally reported by: Ian Mackenzie
- Originally created at: 2013-12-09T17:27:20.992

**Comment by ianmackenzie** [comment]
Visual Studio static analyzer error messages:


```
#!
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Globalization.hpp(57): error C2244: 'opensolid::Globalization<,>::transformationMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Localization.hpp(58): error C2244: 'opensolid::Localization<,>::transformationMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/LinearTransformation.hpp(58): error C2244: 'opensolid::LinearTransformation<>::transformationMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Transplant.hpp(61): error C2244: 'opensolid::Transplant<,>::transformationMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Box.hpp(94): error C2244: 'opensolid::Box<>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Box.hpp(101): error C2244: 'opensolid::Box<>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Box.hpp(199): error C2244: 'opensolid::Box<>::diagonalVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Box.hpp(314): error C2244: 'opensolid::Box<>::operator -' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Box.hpp(321): error C2244: 'opensolid::Box<>::operator -' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Point.hpp(73): error C2244: 'opensolid::Point<>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Point.hpp(80): error C2244: 'opensolid::Point<>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Point.hpp(178): error C2244: 'opensolid::Point<>::operator -' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Point.hpp(185): error C2244: 'opensolid::Point<>::operator -' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(78): error C2244: 'opensolid::CoordinateSystem<,>::{ctor}' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(94): error C2244: 'opensolid::CoordinateSystem<,>::{ctor}' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(111): error C2244: 'opensolid::CoordinateSystem<,>::{ctor}' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(125): error C2244: 'opensolid::CoordinateSystem<,>::basisMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(136): error C2244: 'opensolid::CoordinateSystem<,>::basisVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(143): error C2244: 'opensolid::CoordinateSystem<,>::inverseMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(187): error C2244: 'opensolid::CoordinateSystem<,>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(198): error C2244: 'opensolid::CoordinateSystem<,>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(209): error C2244: 'opensolid::CoordinateSystem<,>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(216): error C2244: 'opensolid::CoordinateSystem<,>::xBasisVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(227): error C2244: 'opensolid::CoordinateSystem<,>::yBasisVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(238): error C2244: 'opensolid::CoordinateSystem<,>::zBasisVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/CoordinateSystem.hpp(248): error C2244: 'opensolid::CoordinateSystem<,>::basisVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Matrix.hpp(156): error C2244: 'Eigen::MatrixBase<Derived>::transplanted' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/Matrix.hpp(169): error C2244: 'Eigen::MatrixBase<Derived>::transplanted' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/LineSegment.hpp(93): error C2244: 'opensolid::LineSegment<>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/LineSegment.hpp(99): error C2244: 'opensolid::LineSegment<>::normalVector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/SpatialCollection/TransformedCollection.hpp(84): error C2244: 'opensolid::detail::TransformedCollection<TBaseCollection,>::transformationMatrix' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/SpatialCollection/TransformedCollection.hpp(180): error C2244: 'opensolid::detail::TransformedCollectionIterator<TBaseCollection,>::{ctor}' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/SpatialCollection/TranslatedCollection.hpp(84): error C2244: 'opensolid::detail::TranslatedCollection<TBaseCollection>::vector' : unable to match function definition to an existing declaration
10>C:\Colony\VS2012\x64\OpenSolid-0.4.6\include\OpenSolid/Core/SpatialCollection/TranslatedCollection.hpp(164): error C2244: 'opensolid::detail::TranslatedCollectionIterator<TBaseCollection>::{ctor}' : unable to match function definition to an existing declaration
```

----------------------------------------
Original comment by: Ian Mackenzie

**Comment by ianmackenzie** [comment]
Should be mostly resolved now with new matrix class hierarchy implemented in revision efd6401.

----------------------------------------
Original comment by: Ian Mackenzie
